app/src/main/java/com/android/unio/model/authentication/AuthViewModel.kt

```diff
@@ -0,0 +1,70 @@
+package com.android.unio.model.authentication
+
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import com.android.unio.model.user.UserRepository
+import com.android.unio.model.user.UserRepositoryFirestore
+import com.android.unio.ui.navigation.Route
+import com.android.unio.ui.navigation.Screen
+import com.google.firebase.Firebase
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.auth.auth
+import com.google.firebase.firestore.firestore
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+class AuthViewModel(
+    private val firebaseAuth: FirebaseAuth,
+    private val userRepository: UserRepository
+) : ViewModel() {
+  private val _authState = MutableStateFlow<String?>(null)
+  val authState: StateFlow<String?>
+    get() = _authState
+
+  init {
+    verifyAuthState()
+  }
+
+  /**
+   * Verifies the authentication state of the user. If the user should be redirected, the
+   * [authState] is updated.
+   */
+  private fun verifyAuthState() {
```

Comment:
Consider renaming this to something along the lines of `addAuthStateVerifier`, to specify that this actually creates a new listener, and is not a one-time check.